File: python/sim.py
import pandas as pd
from datetime import datetime
from bot import Bot, Config
from timewrapper import timeme

# Returns the index of the date specific to the number of desired days in the future
def find_index(dates: list[int], years:int=0) -> int:

    start_date = datetime.fromtimestamp(dates[0])
    if years > 0:
        end_date = start_date.replace(year = start_date.year + years)
    else:
        end_date = datetime.fromtimestamp(dates[len(dates)-1])

    print("Time range: %s - %s" %(start_date.date(), end_date.date()))
    #print("Total Time: %s" % (datetime.fromtimestamp(dates[len(dates)-1]) - datetime.fromtimestamp(dates[0])))

    if years <= 0:
        print("index = %d" % (len(dates)-1))
        return len(dates)-1

    for i, d in enumerate(dates):
        if datetime.fromtimestamp(d).date() == end_date.date():
            print("index = %d" % i)
            return i

@timeme
def run(prices:list[float], dates:list[int], config:Config, verbose:bool = False) -> float:
    bot = Bot(config)

    bot.execute(prices, dates)

    last_price = prices[len(prices)-1]
    total = bot.total_usd + bot.total_btc * last_price

    if verbose:
        print("Gains: $%f" %bot.calculate_gains_usd())
        print("Total USD = $%f" % bot.total_usd)
        print("Current number of BTC=%f with current price $%.2f ~ $%.2f" % (bot.total_btc, last_price, bot.total_btc * last_price))
        print("USD + BTC = $%.2f \n" % (total))

    return total


@timeme
def read_csv(years:int=0):
    df = pd.read_csv('btcalphaUSD.csv')
    dates = df.iloc[:, 0]
    end_index = find_index(dates, years)
    prices = df.iloc[:end_index, 1]
    #print(datetime.fromtimestamp(dates[len(prices)]).date())
    return dates, prices


if __name__ == '__main__':

    dates, prices = read_csv()

    totals = list[int]() 
    price_deviations = list[float]() 
    book = dict[int, Config]()
    book_low_risk = dict[int, Config]()
    i = 0

    #TODO: Search from bigger ranges to smaller ones, diving by 2 each time it's cut (binary search)

    order_array = [10, 15, 20]
    target_profit_perc_array = [0.1, 0.5, 1, 1.5, 2, 2.5, 3, 5, 10]
    price_deviation_safety_orders_array = [0.5, 1, 1.5, 2]
    safety_order_volume_scale_array = [1.05, 1.1, 1.5, 2, 3]
    safety_order_step_scale_array = [0.9, 1, 1.1, 1.2, 1.3]
    # order_array = [10, 15]
    # target_profit_perc_array = [1, 10]
    # price_deviation_safety_orders_array = [0.5, 1]
    # safety_order_volume_scale_array = [1.5]
    # safety_order_step_scale_array = [0.9]

    total_combinations = len(order_array)**2 * len(target_profit_perc_array) * len(price_deviation_safety_orders_array) * len(safety_order_volume_scale_array) * len(safety_order_step_scale_array)

    for base_order_size in order_array:
        for safety_order_size in order_array:
            for target_profit_perc in target_profit_perc_array:
                for price_deviation_safety_orders in price_deviation_safety_orders_array:
                    for safety_order_volume_scale in safety_order_volume_scale_array:
                        for safety_order_step_scale in safety_order_step_scale_array:
                            i += 1
                            config = Config(
                                base_order_size=base_order_size,
                                safety_order_size=safety_order_size,
                                target_profit_perc=target_profit_perc,
                                price_deviation_safety_orders=price_deviation_safety_orders,
                                safety_order_volume_scale=safety_order_volume_scale,
                                safety_order_step_scale=safety_order_step_scale)

                            if i == 2019:
                                config.total = run(prices, dates, config)
                                totals.append(int(config.total))
                                book[int(config.total)] = config
                                price_deviation = config.max_safety_order_price_deviation()
                                book_low_risk[price_deviation] = config
                                price_deviations.append(price_deviation)

                                print("%d/%d : $%.2f" % (i, total_combinations, config.total))

    # order by less risk
    print("\n-------------------------------------")
    for deviation in reversed(sorted(price_deviations, reverse=True)[0:10]):
        print(book_low_risk[deviation])

    # order by most profit
    print("\n-------------------------------------")
    for total in reversed(sorted(totals, reverse=True)[0:10]):
        print(book[total])

    # config = Config(
    #     base_order_size=10.0,
    #     safety_order_size=20.0,
    #     target_profit_perc=1.0,
    #     price_deviation_safety_orders=0.5,
    #     safety_order_volume_scale=2.0,
    #     safety_order_step_scale=1.2)
    # bot = Bot(config, True)

    # _, value =bot.simulate(prices, dates, 0)
    # print(value)
#TODO: for a single simulation I could chart how it bevahes in time

File: go/simulator/sim.go
package simulator

import (
	"fmt"
	"math"
	"sort"
	"time"
)

func Simulate(prices []float64, dates []int64) {
	start := time.Now()

	totals := []int{}
	book := map[int]Config{}
	price_deviations_5 := []int{}
	book_risk_5 := map[int]Config{}
	price_deviations_10 := []int{}
	book_risk_10 := map[int]Config{}
	price_deviations_20 := []int{}
	book_risk_20 := map[int]Config{}
	price_deviations_30 := []int{}
	book_risk_30 := map[int]Config{}
	price_deviations_40 := []int{}
	book_risk_40 := map[int]Config{}
	price_deviations_50 := []int{}
	book_risk_50 := map[int]Config{}
	price_deviations_60 := []int{}
	book_risk_60 := map[int]Config{}
	i := 0

	order_array := createFloatArray(10.0, 20.0, 5.0)
	target_profit_perc_array := createFloatArray(0.1, 2.0, 0.1)
	price_deviation_safety_orders_array := createFloatArray(0.5, 5.0, 0.1)
	safety_order_volume_scale_array := createFloatArray(1.0, 2.0, 0.1)
	safety_order_step_scale_array := createFloatArray(0.8, 1.5, 0.1)

	total_combinations := int(math.Pow(float64(len(order_array)), 2)) * len(target_profit_perc_array) * len(price_deviation_safety_orders_array) * len(safety_order_volume_scale_array) * len(safety_order_step_scale_array)

	for _, base_order_size := range order_array {
		for _, safety_order_size := range order_array {
			for _, target_profit_perc := range target_profit_perc_array {
				for _, price_deviation_safety_orders := range price_deviation_safety_orders_array {
					for _, safety_order_volume_scale := range safety_order_volume_scale_array {
						for _, safety_order_step_scale := range safety_order_step_scale_array {
							i += 1
							bot := Bot{
								Conf: Config{
									BaseOrderSize:              base_order_size,
									SafetyOrderSize:            safety_order_size,
									TargetProfitPerc:           target_profit_perc,
									PriceDeviationSafetyOrders: price_deviation_safety_orders,
									SafetyOrderVolumeScale:     safety_order_volume_scale,
									SafetyOrderStepScale:       safety_order_step_scale,
								},
								Verbose:    false,
								InitialUsd: 1000,
								TotalUsd:   1000,
								TotalBtc:   0,
							}
							bot.Execute(prices, dates)
							totals = append(totals, int(bot.Conf.Total))
							book[int(bot.Conf.Total)] = bot.Conf

							fmt.Printf("%d/%d : $%.2f\n", i, total_combinations, bot.Conf.Total)
						}
					}
				}
			}
		}
	}

	sort.Ints(totals)

	for _, total := range totals {
		a := book[total]
		price_deviation := a.MaxSafetyOrderPriceDeviation()
		if price_deviation > 4 && price_deviation < 6 {
			key := price_deviation * a.Total
			price_deviations_5 = append(price_deviations_5, int(key))
			book_risk_5[int(key)] = a
		} else if price_deviation > 9 && price_deviation < 11 {
			key := price_deviation * a.Total
			price_deviations_10 = append(price_deviations_10, int(key))
			book_risk_10[int(key)] = a
		} else if price_deviation > 19 && price_deviation < 21 {
			key := price_deviation * a.Total
			price_deviations_20 = append(price_deviations_20, int(key))
			book_risk_20[int(key)] = a
		} else if price_deviation > 29 && price_deviation < 31 {
			key := price_deviation * a.Total
			price_deviations_30 = append(price_deviations_30, int(key))
			book_risk_30[int(key)] = a
		} else if price_deviation > 39 && price_deviation < 41 {
			key := price_deviation * a.Total
			price_deviations_40 = append(price_deviations_40, int(key))
			book_risk_40[int(key)] = a
		} else if price_deviation > 49 && price_deviation < 51 {
			key := price_deviation * a.Total
			price_deviations_50 = append(price_deviations_50, int(key))
			book_risk_50[int(key)] = a
		} else if price_deviation > 59 && price_deviation < 61 {
			key := price_deviation * a.Total
			price_deviations_60 = append(price_deviations_60, int(key))
			book_risk_60[int(key)] = a
		}
	}

	if len(price_deviations_5) > 0 {
		fmt.Println("\nPrice Deviation around 5%")
		a := book_risk_5[price_deviations_5[len(price_deviations_5)-3]]
		fmt.Print(a.ToString())
	}
	if len(price_deviations_10) > 0 {
		fmt.Println("\nPrice Deviation around 10%")
		a := book_risk_10[price_deviations_10[len(price_deviations_10)-3]]
		fmt.Print(a.ToString())
	}
	if len(price_deviations_20) > 0 {
		fmt.Println("\nPrice Deviation around 20%")
		a := book_risk_20[price_deviations_20[len(price_deviations_20)-3]]
		fmt.Print(a.ToString())
	}
	if len(price_deviations_30) > 0 {
		fmt.Println("\nPrice Deviation around 30%")
		a := book_risk_30[price_deviations_30[len(price_deviations_30)-3]]
		fmt.Print(a.ToString())
	}
	if len(price_deviations_40) > 0 {
		fmt.Println("\nPrice Deviation around 40%")
		a := book_risk_40[price_deviations_40[len(price_deviations_40)-3]]
		fmt.Print(a.ToString())
	}
	if len(price_deviations_50) > 0 {
		fmt.Println("\nPrice Deviation around 50%")
		a := book_risk_50[price_deviations_50[len(price_deviations_50)-3]]
		fmt.Print(a.ToString())
	}
	if len(price_deviations_60) > 0 {
		fmt.Println("\nPrice Deviation around 60%")
		a := book_risk_60[price_deviations_60[len(price_deviations_60)-3]]
		fmt.Print(a.ToString())
	}

	fmt.Println("\nBest performers")
	totals = totals[len(totals)-10:]
	for _, total := range totals {
		a := book[total]
		fmt.Print(a.ToString())
	}

	fmt.Println("\nSimulation execution duration: ", time.Since(start))
}

func createFloatArray(start, end, step float64) []float64 {
	array := []float64{}
	for i := start; i <= end; i += step {
		array = append(array, float64(i))
	}

	return array
}
